---
title: "coSMicQC: A step toward quality control and improvement of morphology profiles"
author: Jenna Tomkinson
member: jenna-tomkinson
tags:
  - opinions
  - process
  - software
---
# Let's tour the single-cell morphological galaxy with coSMicQC!

Hi everyone! 🤩

Wow, it has been over a year since the last blog post.
*cough*
I think some dust might have accumulated. 🧹

All jokes aside, it has been a very busy year for the Way Lab!
We have posted three preprints:

1. [Reproducible image-based profiling with Pycytominer](https://pmc.ncbi.nlm.nih.gov/articles/PMC10690292/)
2. [A morphology and secretome map of pyroptosis](https://www.biorxiv.org/content/10.1101/2024.04.26.591386v1)
3. [High-content microscopy and machine learning characterize a cell morphology signature of NF1 genotype in Schwann cells](https://www.biorxiv.org/content/10.1101/2024.09.11.612546v1)

We have also published one paper just this month, called [Toward generalizable phenotype prediction from single-cell morphology representations](https://bmcmethods.biomedcentral.com/articles/10.1186/s44330-024-00014-3). 🎉

Now that I have a bit of bandwidth, I wanted to take the time to come back to go over something that I am very excited and passionate about:

✨ *Quality control of single-cell morphology profiles* ✨

In this blog post, I will go over what is quality control for morphology profiles, what current work exists on this topic, the software (coSMicQC) that we have developed to help in this topic, and where we hope to go from here.

## Contents

1. 📋 [What is quality control for single-cell morphology profiles?](https://www.waysciencelab.com/2024/12/19/cosmicqc#defining-quality-control-of-single-cell-morphology-profiles)
2. 📉 [What current exists for this topic?](https://www.waysciencelab.com/2024/12/19/cosmicqc#what-currently-exists)
3. 🌌 [Intro to coSMicQC](https://www.waysciencelab.com/2024/12/19/cosmicqc#introduction-to-cosmicqc)
4. 🎯 [Example of impact](https://www.waysciencelab.com/2024/12/19/cosmicqc#example-of-cosmicqc-impact)
5. 🚀 [Where do we go from here?](https://www.waysciencelab.com/2024/12/19/cosmicqc#whats-next)

## Defining quality control of single cell morphology profiles

> **Quality Control (QC)**  
> The process of validating and enforcing a specific standard across all products.

> **Single-cell morphology profiles**  
> Features (e.g., area, texture, intensity, etc.) extracted from images of cells. 
> Traditionally, profiles are formatted in a data frame where each row is single-cell, and columns are either metadata or feature extracted from a specific segmented part of a cell (e.g., nucleus, whole cell, or cytoplasm). 

Putting these two definitions together, the main definition is the process of enforcing a standard quality across the single cells and their features derived from their respective segmented parts.

### Background on image-based profiling

If you are unfamiliar to the standard process of image-based profiling, there are typically 2 main steps after collecting raw images:

{%
  include figure.html
  image="images/blog/cosmicqc/simple_image_based_profiling.png"
  width="100%"
%}
> Simple workflow for image-based profiling including intermediate steps for CellProfiler and pycytominer.

Extraction of morphology features from images is traditional performed using the software [CellProfiler](https://cellprofiler.org/). 
There are other feature extraction softwares like incarta, but we will be focusing on CellProfiler in this blog post. 
CellProfiler is used to perform multiple tasks, include illumination correction, segmentation (traditionally of nuclei, cells, and cytoplasm), and feature extraction. 

**PLUG TIME!**
If you aren't familiar with illumination correction (IC) or want to learn more, feel free to read the two blog posts on the topic, "[Illumination Correction Made Easier](https://www.waysciencelab.com/2022/08/09/illumcorrect.html)" and "[Steps for Performing Illumination Correction in Microscopy Images](https://www.waysciencelab.com/2023/08/07/illumsteps.html)". 
I also created a blog post regarding the pros and cons of two segmentation methods if you are interested called "[Segmentation Software Usability and Performance: Part I](https://www.waysciencelab.com/2022/11/01/segmentation.html)".
**OKAY, PLUG TIME OVER!**

Once features are extracted, the current most popular method of processing is by aggregating the single-cell features to well-level population, either through median or averaging.
This process can be done within a software called, [pycytominer](https://pycytominer.readthedocs.io/en/stable/), which is optimized to process morphological profiles from multiple outputs.
Aggregation removes the heterogeneity of the population and reduces noise.
But, this "reduction of noise" is a bit fishy to me.
If you aggregate a bunch of poorly segmented single-cell profiles, then aggregating will just exacerbate the issue and mis-represent the biology of the population.
This leads me into why I think quality control at the single-cell level for morphology profiles is a necessary step, regardless of if you plan to process your data as bulk (aggregated) or single cells.

### Why single-cell quality control?

Quality control of single-cell profiles is very important to avoid mis-interpretation of the results, and ensure that downstream analysis is picking up biologically relevant information.
Imagine you have an picture of cells and you want to cut out each individual cell.
As you are cutting, your hand could end up slipping or you get distracted, so your individual cell cut-out ends up missing part of the cell.
This is what a segmentation error is. 
These cut-outs with incomplete cells are missing vital and biological information about that cell.
Quality control acts like a checklist, ensuring that each cut-out meets the required standards, such as verifying that no parts of the cell are missing.

{%
  include figure.html
  image="images/blog/cosmicqc/segmentation_cartoon.png"
  width="100%"
%}
> *Simplified representation of single-cell segmentation.*

In the Way Lab, we traditionally process single-cell data, so we expect heterogeneity within our samples.
We do not want the heterogeneity influenced by technical factors, like poor segmentations.
So, we want to remove any of these poor segmentation before we perform important pre-processing steps like normalization and feature selection, which will be influenced noise.
The Way Lab creates and maintains softwares, including pycytominer and [CytoTable](https://cytomining.github.io/CytoTable/) (which is not including in the above example, but we expect to become a standard in the future).
CytoTable reformats the output from CellProfiler, where each feature per segmentation is in its own table, to one parquet file each row is a single-cell and the columns are the combined features per cell from each table.

Single-cell quality control for morphology profiles should occur in between CytoTable and pycytominer.
But, what methodology exists so that we can perform this task?

---

## What currently exists?

If you are familiar with the RNA-seq world, you will know that the concept of *Single-cell quality control* itself is not new.
It is a well known process within scRNA-seq pipelines where individual cells are filtered out using a metric(s) that detects if it is likely an artifact or contains noise.
It is such a well-established part of this field that there even is [a whole website detected to scRNA tools](https://www.scrna-tools.org/), where can filter for specifically [quality control tools](https://www.scrna-tools.org/tools?sort=name&cats=QualityControl), which provides at least one hundred different options to try!
This is also an important step in not just single-cell, but also bulk RNA-seq as well, which has existed for over two decades.
scRNA-seq has been around for over a decade, so there has been a lot of time to develop these methods across both technologies.

[Cell Painting](https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0080999) was first developed in 2013, meaning that image-based profiling and extraction of morphological profiles is a new field, but still been around are 11 years.
Single-cell quality control for morphology profiles is not has well developed in the image-based profiling realm.

I will preface this by saying that you are currently able to filter objects in CellProfiler using morphology measurements, but the module isn't very intuitive.
You can set limits for what measurement is considered good or bad, but how would you know what the limits of a feature is without exporting the data and performing analyses to get a sense of the range of values for your dataset.
In my opinion, this module is good to have, but it isn't a solution to single-cell quality control.

I did conduct some research into seeing if there were already methods out there to perform this task.
I was only able to identify one paper that makes an attempt at this process.
The publication in BMC Bioinformatics is called "[A cell-level quality control workflow for high-throughput image analysis](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-020-03603-5)", and it was published back in 2020.
In the background section, they state:

> *"To our knowledge, this represents the first QC method and scoring metric that operate at the cell level."*

Since I couldn't find any other paper that focuses on quality control at the single-cell level, I completely agree with their statement. 
To summarize the paper, they present a cell-level quality control workflow which uses machine learning to detect artifacts within images.
They utilized CellProfiler to segment single cells and extract features from both the cells and whole images.
They wanted to group images based on their phenotype, so they utilized CellProfiler features (like `PowerLogLogSlope`, `FocusScore`, `MeanIntensity`, `Cell Count`, etc.) to detect images that contains artifacts versus no artifacts.
Once images were grouped together, they performed a method using KDE to sample the images into their respective phenotypes.
They then trained an SVM (Support Vector Machine) per phenotype group, to classify where there are artifacts and where there are cells.
I believe I barely scratch the surface of the complexities behind this method, so feel free to read the paper if you want a more in-depth look.

What I really liked about this paper is that it is attempting to preserve images that include artifact, as these images still contain other cells.
It is different than the current process where whole images are removed that fail QC metrics like blur and over-saturation (traditional done in CellProfiler).
But, I also have my suspicions about using cells that come from an image with corruption, such as an artifact.
I wonder how much things like an artifact or smudge could be directly impacting the biology the cells.
In my opinion, I would rather default to losing those cells in an image than risk any issues in downstream interpretation of the results do to any impacts the artifacts in an image had.

I have two main concerns from this paper, specifically for it's generalizability and usability in my own pipelines:

1. It was not clear to me how this method is filtering out artifacts at the single-cell level. From what I understood from the figures, the SVM are able to detect where there is artifacts and where there are cells, but I didn't see where in their pipeline to they remove segmentations from the CellProfiler output. This is a big concern for generalizability as it seems like you would need to train SVMs on your data to be able to apply this method but there isn't a direct link to that output and filtering cells.
2. The code is available via FigShare, but it was a bit hard to find and parse. Since it is not on GitHub, that also limits the usability since there isn't any environment or dependencies I could find and there isn't a way to create issues and contact the maintainers for assistance. As well, I found that the code I could find was not easy to understand and lacked documentation. I want to make clear, I appreciate that they made the code available and open source which is more than some other papers. My main point is that I don't think this method is feasible to include in my standard workflow.

In conclusion, I want to thank the authors of this paper for their hard work and their important contributions to pushing the field towards developing quality control methodology at the single-cell level for morphology profiles.
This paper fueled the fire that I had started in myself about wanting to pursue working on a new methodology, one that is simple, to filter out poor quality segmentations from single-cell morphology profiles.

---

## Introduction to coSMicQC

After 2 and half years of being in my position and performing image-based profiling, I have come to realize that I am not perfect (shocking I know! 😱).
No matter how much I try to manually optimize my segmentation parameters in CellProfiler or when using [Cellpose](https://www.cellpose.org/) or [Stardist](https://stardist.net/), it will not work for all of the single cells in my datasets.
Now, I am not saying this is a fault of my own, but segmentation errors are a persistent issue and is currently a problem that hasn't been solved to this day.

My point is, you can expect in all of your datasets that it will contain some poorly segmented cells that can be described as technical artifacts.
You can also expect that debris or smudges could also end up being mis-segmented as cells.
When working with single-cell data, we do not want these technical artifacts impacting the interpretation of our results.
Thus, was born this simple idea in my head.

Why not just filter out single cells based on the extracted morphology features?

You wouldn't need to train any model or perform any other pipeline.
You already have the data in your lap to use, but it is how you use that data is important.

In comes, [coSMicQC](https://github.com/WayScience/coSMicQC), or Single cell Morphology Quality Control!

This new method works by creating conditions, represented by dictionaries.
A condition is a feature or group of features that will detect a specific type of technical artifact.
For example, nuclei area and intensity can work together as a condition to detect large nuclei with high intensity which represents segmentation of clustered/overlapping nuclei.
You condition will also need to include the standard deviation from the mean in which you expect these outliers to be located.
If you want to detect cells above a mean, you use a positive value, and vice versa for cells below the mean.

{%
  include figure.html
  image="images/blog/cosmicqc/area_outlier_example.png"
  width="100%"
%}
> Example of how outliers are detected based on nuclei area in coSMicQC.

We then apply the function `find_outliers` to the simple conditions independently, yielding a dataframe with only the identified outliers, including metadata and the features used to detect the outliers.
A summary is also outputted, which includes the total number of outliers detected, the percent of the total cells in the data detected as outliers, and the range of the values of the outliers for the features used in the condition.

A highlight of coSMicQC is that you are able to visualize single-cell segmentations within the notebook if you provide the original images and exported outlines from CellProfiler, through the use of a dataframe format called [CytoDataFrame](https://github.com/WayScience/CytoDataFrame).
This makes it more streamlined to assess if your conditions are working without having to go outside of the notebook via your file manager, Fiji, or CellProfiler to see what the segmented cells look like.
This feature is still being developed as we find more datasets with varying conditions for image file structure.

Once you have optimized the conditions, you can then use the function `label_outliers` which will output the original dataframe with columns for each condition and if a single-cell passed or failed each condition.
You can then choose to either save the original dataframe with the new columns and keep outliers or remove single cells that were detected as outliers.
You can also not use the coSMicQC functionality and take the indices from the `find_outliers` outputs and remove single cells that way.
We have elected to call the outputs that come from coSMicQC as `cleaned` profiles.

With these cleaned profiles, further downstream processes should yield higher quality results.
We have been able to test this hypothesis using one of our projects, and prove that there is merit to pursing single-cell quality control. 

---

## Example of coSMicQC impact

The conception of coSMicQC came from this project, which unfortunately is now within a private repository (due to external factors).
The repository, titled [Cell Painting predicts cardiac fibrosis](https://github.com/WayScience/cellpainting_predicts_cardiac_fibrosis), will become public once the respective manuscript has been published.

A brief overview of the project is as follows:

Cardiac fibroblasts from tissue of both heart failure patients and organ donor patients without heart failure were collected, plated, and imaged using a modified Cell Painting assay. 
We performed an image-based profiling workflow to segment and extract single-cell morphology features. 
We trained a logistic regression binary classifier to predict if a single-cell comes from a failing or non-failing heart.

Within the workflow, we performed single-cell quality control to clean the data using coSMicQC.
One of the biggest problems in this dataset is the high confluence, which occurs due to proliferation.
This leads to very clustered nuclei and higher intensity of the cells across organelles within the clusters compared to the peripheral cells.
We created two conditions to find these exact poor segmentation: cells area to find abnormally small cells and large nuclei area and high nuclei intensity to detect over-segmented nuclei clusters.

{%
  include figure.html
  image="images/blog/cosmicqc/conditions_example_plot.png"
  width="100%"
%}
> Plots of conditions with example FOVs of good versus failed single cells.

To evaluate if single-cell quality control is important to training a high quality model, we perform a bootstrapping analysis using holdout data (e.g., data that the model never saw during training).
We trained a new model with the same parameters, but using non-quality controlled training data.
There are two holdout datasets to evaluate, one that has had QC performed on the data and another without.
We will apply the models to their respective datasets, e.g., QC model with QC holdout data and no QC model with no QC holdout data.

We then perform what is called bootstrapping, where sub-samples of the data are taken to collect 1,000 different datasets of the same size as the original dataset.
This means that it will randomly take single cells from the data (sometimes repeated) to create a new population, which represents what we can expect in real-life.
We then calculate the ROC AUC metric, or  Receiver Operating Characteristic Area Under the Curve.
To be specific, this is measuring the area under the curve for the ROC, which is a plot that shows the trade off between the true positive rate and false positive rate, where a good classifier will have an area under the curve closer to 1 and bad classifier will be closer to 0.
We measure the ROC AUC for all 1,000 subsample to create a distribution, which we then plot as a histogram.
The mean of the distributions are represented as dashed lines.

{%
  include figure.html
  image="images/blog/cosmicqc/histogram_roc_auc.png"
  width="100%"
%}
> Histogram of the ROC AUC metrics comparing the results from the QC model applied to QC data and no QC model applied to no QC data.

When we apply QC to our data (orange), we show a statistically significant improvement in performance at classification (t-stat = -103.7, p-value = 0.0) compared to not applying QC at all (blue).
By removing poor quality segmentation, we reduce the noise in the data and significantly improve the performance of our model when applied to new data (increased generalizability).

The implementation of coSMicQC was very simple for this project.
The hardest part was picking which features to use as conditions.
It only takes less than a minute to run coSMicQC in a notebook to detect outliers and save the cleaned data.
This project is the first of hopefully many examples where we can show the importance of single-cell quality control!

---

## What's next?

We have already implemented this multiple other projects!
Now, I am not going to sit here writing this blog and pretend like coSMicQC works without any fault.
I want to be very clear, this method is very good when applying it to cells without perturbation (e.g., happy, normal cells).
We applying coSMicQC to a dataset where there are cells treated with many different perturbations, we saw that nuclei with very interesting phenotypes ended up failing quality control (red) at a higher rate than those that looked more normal.

{%
  include figure.html
  image="images/blog/cosmicqc/phenotype_example.png"
  width="100%"
%}
> Example of passing (green) and failing (red) single cells for a normal phenotype FOV (left) and interesting phenotype FOV (right).

This means that unfortunately, coSMicQC is not robust to some phenotypes and might over-correct the data.
But, quality control is not a perfect science.
The purpose of quality control, regardless of field, is to maximize the amount of poor quality cell removed while minimizing the amount of good quality cells being wrongly eliminated from the dataset.
This means that no matter what, you must expect some false positives, but we can attempt to control for this.

Given that we know the current limitations of the current simple methodology of coSMicQC, we are looking to improve by finding new methods to implement into our software.
Some of the current ideas we are thinking of testing include DBSCAN (proposed by Erik Serrano) and PCA or UMAP to utilize either all features or selected features to find outliers based on clusters.

Our big goal for 2025 is publish a paper for coSMicQC, which means we will be testing on more datasets.
We plan on updating this blog with more sections that include the results from evaluating on more data, so please look out for more!

---

## Final thoughts

If you have gotten to the end of this blog and have an idea that you want to propose, please feel free to [add an issue](https://github.com/WayScience/coSMicQC/issues) to the coSMicQC GitHub!
We would really like to make a push towards standardizing this process within the traditional image-based profiling workflow.
We hope to build a community around single-cell quality control for morphology profiles, so we can lead the field to a future with higher-quality datasets.

Thank you for reading, and happy profiling! 👩🏻‍💻
